core/src/main/java/ai/timefold/solver/core/impl/solver/AssignmentProcessor.java

```diff
@@ -0,0 +1,158 @@
+        entityPlacer.solvingStarted(solverScope);
+        entityPlacer.phaseStarted(phaseScope);
+        entityPlacer.stepStarted(stepScope);
+
+        try (scoreDirector) {
+            var placementIterator = entityPlacer.iterator();
+            if (!placementIterator.hasNext()) {
+                throw new IllegalStateException("""
+                        Impossible state: entity placer (%s) has no placements.
+                        """.formatted(entityPlacer));
+            }
+            var placement = placementIterator.next();
+            var recommendedAssignmentList = new ArrayList<Recommendation_>();
+            var moveIndex = 0L;
+            for (var move : placement) {
+                recommendedAssignmentList.add(execute(scoreDirector, move, moveIndex, clonedElement, valueResultFunction));
+                moveIndex++;
+            }
+            recommendedAssignmentList.sort(null);
+            return recommendedAssignmentList;
+        } finally {
+            entityPlacer.stepEnded(stepScope);
+            entityPlacer.phaseEnded(phaseScope);
+            entityPlacer.solvingEnded(solverScope);
+        }
+    }
+
+    private EntityPlacer<Solution_> buildEntityPlacer() {
+        var solver = (DefaultSolver<Solution_>) solverFactory.buildSolver();
+        var phaseList = solver.getPhaseList();
+        var constructionHeuristicCount = phaseList.stream()
+                .filter(DefaultConstructionHeuristicPhase.class::isInstance)
+                .count();
+        if (constructionHeuristicCount != 1) {
+            throw new IllegalStateException(
+                    "Assignment Recommendation API requires the solver config to have exactly one construction heuristic phase, but it has (%s) instead."
+                            .formatted(constructionHeuristicCount));
+        }
+        var phase = phaseList.get(0);
+        if (phase instanceof DefaultConstructionHeuristicPhase<Solution_> constructionHeuristicPhase) {
+            return constructionHeuristicPhase.getEntityPlacer();
+        } else {
+            throw new IllegalStateException(
+                    "Assignment Recommendation API requires the first solver phase (%s) in the solver config to be a construction heuristic."
+                            .formatted(phase));
+        }
+    }
+
+    private Recommendation_ execute(InnerScoreDirector<Solution_, Score_> scoreDirector, Move<Solution_> move, long moveIndex,
+            In_ clonedElement, Function<In_, Out_> propositionFunction) {
+        var undo = move.doMove(scoreDirector);
+        var newScoreAnalysis = scoreDirector.buildScoreAnalysis(fetchPolicy == ScoreAnalysisFetchPolicy.FETCH_ALL);
+        var newScoreDifference = newScoreAnalysis.diff(originalScoreAnalysis);
+        var result = propositionFunction.apply(clonedElement);
+        var recommendation = recommendationConstructor.apply(moveIndex, result, newScoreDifference);
```

Comment:
I understand what the logic is doing here and have some questions:

1. Wouldn't this create a capturing lambda?
2. We will return `DefaultRecommendedAssignment` or `DefaultRecommendedFit`. What is the advantage of using the lambda instead of just instantiating the default implementations?

Reply:
Wrt. 1 - no. The lambda is already created and passed as an argument. Also, all of the variables are coming as arguments, nothing is captured externally. 

Wrt. 2 - assignment and fit have no common ancestor. Therefore it would be impossible to have the same class handle both assignments and fits. Doing it like this allows me to reuse 100 % of the code.